libsolidity/codegen/ir/IRLValue.cpp

```diff
@@ -166,3 +166,75 @@ string IRStorageArrayLength::setToZero() const
 {
 	return storeValue("0", *TypeProvider::uint256());
 }
+
+IRMemoryItem::IRMemoryItem(
+	IRGenerationContext& _context,
+	std::string _address,
+	bool _calldata,
+	bool _padded,
+	Type const& _type
+):
+	IRLValue(_context, &_type),
+	m_address(move(_address)),
+	m_padded(_padded),
+	m_calldata(_calldata)
+
+{ }
+
+string IRMemoryItem::retrieveValue() const
+{
+	if (m_type->isValueType())
+		return m_context.utils().readFromMemory(*m_type, m_calldata, m_padded) +
+			"(" +
+			m_address +
+			")\n";
+	else if (m_calldata)
```

Comment:
I think it is more complicated than that (you have to add the base offset), better to leave it out for now.

Reply:
But that is already done before we pass the value to IRMemoryItem:
```
			case DataLocation::CallData:
			{
				string const memAddress =
					(arrayType.location() == DataLocation::Memory ?
						m_utils.memoryArrayIndexAccessFunction(arrayType) :
						m_utils.calldataArrayIndexAccessFunction(arrayType)) +
					"(" +
					m_context.variable(_indexAccess.baseExpression()) +
					", " +
					m_context.variable(*_indexAccess.indexExpression()) +
					")";

				setLValue(_indexAccess, make_unique<IRMemoryItem>(
					m_context,
					memAddress,
					arrayType.location() == DataLocation::CallData,
					true,
					*arrayType.baseType()
				));
				break;
			}
```